### ISSUE TITLE
RouterLink produces a different parameter to UI#navigate

### ISSUE THREAD

**Issue by eriklumme** [vaadin/flow#5275]
Consider the following two ways of navigating to a view that has a string parameter

    String parameter = "åäö";

    RouterLink routerLink = new RouterLink("Router link", RouteView.class, parameter);

    Button buttonLink = new Button("Button link");
    buttonLink.addClickListener(e -> UI.getCurrent().navigate(RouteView.class, parameter));

### Expected result

Both produces the same parameter in the `RouteView#setParameter` method

### Actual result

The RouterLink produces `%C3%A5%C3%A4%C3%B6`, while UI#navigate produces `åäö`

### Versions

Platform 13.0.0
Chrome 72.0.3626.121

**Comment by denis-anisimov** [comment]
What do you mean "produces" ?

`%C3%A5%C3%A4%C3%B6` looks like encoded URI _by a browser_.
`UI:: navigate`  doesn't go through browser so it's not encoded.

I don't see an obvious reason why the URI in the browser should be the same as the string used on the server side.
Or you mean `routerLink` has it on the server side ?


I would say the issue is: you should take care about encoding of parameters by yourself with router.
And you have to do it right now.

Different work on the server side is a bug.
Automatic encoding absence is an enhancement: at the moment the workaround is to encode it in the application (though I would definitely avoid using parameters which required encoding at all).

**Comment by eriklumme** [comment]
I think the core issue it the same as in https://github.com/vaadin/flow/issues/5274

When creating a `RouterLink`, or adding the `router-link` attribute to an anchor tag, the navigation event when clicking will be intercepted by the client side `RouterLinkHandler`, a navigation event will be sent to the server, and the event's default behavior will be prevented.

I believe the `RouterLinkHandler`'s use of `getHref()` returns the same result as `someAnchor.href`, which at least in Firefox and Chrome returns an URL-encoded href.

--

I am not sure about the best solution to this. The `location` property of the navigation events could always be decoded server-side by Flow, but I'm not sure if there are edge cases where that wouldn't work either.

**Comment by denis-anisimov** [comment]
I didn't understand the answer to my question: `What do you mean "produces" ?`

>I think the core issue it the same as

I don't see an issue at all at the moment. Please clarify it.
Do you get _something_ (I would like to understand what) on the server side in both cases or in one case you are talking about server side behavior and in other case you are talking about client side ?

`buttonLink.addClickListener(e -> UI.getCurrent().navigate(RouteView.class, parameter));` doesn't produce anything. It does navigation.

`new RouterLink("Router link", RouteView.class, parameter);` makes a router link element in HTML.

**Comment by eriklumme** [comment]
I mean that upon navigation, a value is assigned to the `parameter` parameter of the `setParameter` method.

If this is not an issue, then I don't see how the other one can possibly be. But I believe both are.

From a server-side developer standpoint, I would expect that I can pass a parameter to a RouterLink the same way I can pass it to UI#navigate, and with the same end result.

**Comment by eriklumme** [comment]
> Do you get something on the server side in both cases

Yes. `setParameter` is called on my view both times, in the first case with the parameter `%C3%A5%C3%A4%C3%B6`, in the second case with `åäö`.

**Comment by denis-anisimov** [comment]
Where is the `setParameter` method call in your description ?

**Comment by denis-anisimov** [comment]
Let's go that direction because I see huge confusion:

please provide a source code which I can use to copy paste with the comments on the line where the result is unexpected (and the expected result).

**Comment by eriklumme** [comment]
Apologies for the confusion

    @Route("routes")
    public class RouteView extends VerticalLayout implements HasUrlParameter<String> {
    
        public RouteView() {
            super();
    
            String parameter = "åäö";
    
            RouterLink routerLink = new RouterLink("Router link", RouteView.class, parameter);
    
            Button buttonLink = new Button("Button link");
            buttonLink.addClickListener(e -> UI.getCurrent().navigate(RouteView.class, parameter));
    
            add(routerLink, buttonLink);
        }
    
        @Override
        public void setParameter(BeforeEvent event, @OptionalParameter String parameter) {
            // I expect this to be "åäö" in both cases, it is %C3%A5%C3%A4%C3%B6 in the first one
            System.out.println(parameter);
        }
    }

**Comment by denis-anisimov** [comment]
Very good.
Thanks, now it's pretty clear.

**Comment by davidef** [comment]
Did someone managed to find a way to workaround this ? To me it seams to have a major impact not a minor one (expect if no one is using HasUrlParameter with Strings).

The browser show the same url both using the router link or button navigation but in setParameter we get different values in one case an url-encoded string and in the other we got already decoded string. How can we know if we have to decode or not the received parameter ? I think we should received decoded values only.

**Comment by Artur-** [comment]
Can't reproduce this. A link to a view like
```java
RouterLink specialLinkWithParam = new RouterLink("Link with parameter åäö", ViewWithParam.class, "åäö");
```
results in
```html
<a router-link="" href="paramview/%C3%A5%C3%A4%C3%B6">Link with parameter åäö</a>
```

in the browser and when clicking, `setParameter` is called with `åäö`

When using
```java
UI.getCurrent().navigate(ViewWithParam.class, "åäö");
```
in a click listener, then `setParameter` is also called with `åäö`

**Comment by davidef** [comment]
We are running this with 23.2.x and you can see the issue with the example below. It seams already fixed in 23.3.1+ (seams this: https://github.com/vaadin/flow/pull/15414 ). I think I'll try to upgrade soon to 23.3.x so we get the fix. The other apps still on 14 do not have HasUrlParameter with special characters so I don't know if is reproducible there.

```
@PageTitle("Main")
@Route(value = "")
public class MainView extends VerticalLayout implements HasUrlParameter<String> {
    private static final long serialVersionUID = 1L;

    public MainView() {
       Button button = new Button("test UI.getCurrent().navigate", e -> {
    	   UI.getCurrent().navigate(MainView.class, "url encoded");
       });
       add(button);
       RouterLink link = new RouterLink("test RouterLink", MainView.class, "url encoded");
       add(link);
    }
    
	@Override
	public void setParameter(BeforeEvent event, @OptionalParameter String parameter) {
		Notification.show("Trigger: " + event.getTrigger() + " Param: " + parameter);
	}
}
```
